compiler/fir/entrypoint/src/org/jetbrains/kotlin/fir/session/FirJvmOverridesBackwardCompatibilityHelper.kt

```diff
@@ -0,0 +1,78 @@
+
+/**
+ * Helper that determines if `override` keyword can be omitted for certain overrides. In general it can be omitted if
+ *
+ * 1. the super member is annotated with `@kotlin.internal.PlatformDependent`
+ * 2. the super member is declared in a Java class that has special Kotlin class mapping. For example, overriding members of
+ *    `kotlin.Throwable` does not require `override` keyword.
+ *
+ * Note that, in case of multi-override, if any super member requires `override`, then the `override` keyword cannot be omitted.
+ */
+object FirJvmOverridesBackwardCompatibilityHelper : FirOverridesBackwardCompatibilityHelper {
+    private val javaOrigin = setOf(FirDeclarationOrigin.Java, FirDeclarationOrigin.Enhancement)
+    private val platformDependentAnnotation = ClassId.fromString("kotlin/internal/PlatformDependent")
+
+    override fun overrideCanBeOmitted(
+        overriddenMemberSymbols: List<FirCallableSymbol<*>>,
+        context: CheckerContext
+    ): Boolean {
+        // Members could share the same common interface up in the hierarchy. Hence we track the visited members to avoid redundant work.
+        val visitedSymbols = hashSetOf<FirCallableSymbol<*>>()
+        return overriddenMemberSymbols.all { isPlatformSpecificSymbolThatCanBeImplicitlyOverridden(it, visitedSymbols, context) }
+    }
+
+    private fun isPlatformSpecificSymbolThatCanBeImplicitlyOverridden(
+        overriddenSymbol: FirCallableSymbol<*>,
+        visitedSymbols: MutableSet<FirCallableSymbol<*>>,
+        context: CheckerContext
+    ): Boolean {
+        val fir = overriddenSymbol.fir as? FirCallableMemberDeclaration<*> ?: return false
+        if (fir.isFinal) return false
+
+        if (overriddenSymbol in visitedSymbols) return true
+        visitedSymbols += overriddenSymbol
+
+        val originalMember = fir.originalOrSelf()
+        if (originalMember.annotations.any { it.annotationTypeRef.coneTypeSafe<ConeClassLikeType>()?.classId == platformDependentAnnotation }) {
+            return true
+        }
+
+        if (originalMember.origin !in javaOrigin) return false
+        val containingClassName = originalMember.containingClass()?.classId?.asSingleFqName()?.toUnsafe() ?: return false
+        // If the super class is mapped to a Kotlin built-in class, then we don't require `override` keyword.
+        if (JavaToKotlinClassMap.mapKotlinToJava(containingClassName) != null) return true
+
+        val scope =
+            overriddenSymbol.dispatchReceiverTypeOrNull()?.toRegularClass(context.session)?.unsubstitutedScope(context) ?: return false
+        return when (originalMember) {
+            is FirSimpleFunction -> scope.getDirectOverriddenFunctions(originalMember.symbol)
+            is FirProperty -> scope.getDirectOverriddenProperties(originalMember.symbol)
+            else -> return false
+        }.run {
+            isNotEmpty() && all { isPlatformSpecificSymbolThatCanBeImplicitlyOverridden(it, visitedSymbols, context) }
```

Comment:
I think this will be clearer if you separate out the "not empty" condition. e.g.,
```kt
val overriddenSymbols = when (originalMember) { ... }
return if (overriddenSymbols.isEmpty()) {
  // This symbol requires `override` since it didn't match any of the criteria above
  false
} else {
  overriddenSymbols.all { isPlatformSpecificSymbolThatCanBeImplicitlyOverridden(it, visitedSymbols, context) }
}
```

and I would also rename `overriddenSymbol` in the function parameters to `symbol` to avoid confusion.

Reply:
Sounds good!